### PR TITLE
Correctly get binder when a local function is the first thing in top level statements

### PR DESCRIPTION
**Pull request by 333fred**
We had a bug where nullable analysis was getting the incorrect binder when attempting to populate initial information about all top-level statements if the first thing in the top level was a local function. This ensures that we use the correct binder for these cases.

I've investigated other areas where we call `GetAdjustedNodePosition`, and I do not believe that any of them have this issue. It seems like this particular case is unique in that preferring the innermost binder was actually not what we wanted to do here, or the usage wasn't possible to hit with this scenario.

Fixes https://github.com/dotnet/roslyn/issues/47467

### CODE REVIEW

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler for reviews please

File: src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs [hunk 1]
```diff
@@ -1975,7 +1975,24 @@ protected void EnsureNullabilityAnalysisPerformedIfNecessary()
 
             BoundNode bind(CSharpSyntaxNode root, DiagnosticBag diagnosticBag, out Binder binder)
             {
-                binder = GetEnclosingBinder(GetAdjustedNodePosition(root));
+                if (root is CompilationUnitSyntax)
+                {
+                    // Top level statements are unique among our nodes: if there are no syntax nodes before local functions,
+                    // then that means the start of the span of the top-level statement is the same as the start as the local
+                    // function. Therefore, GetEnclosingBinder can't tell the difference, and it will get the binder for the
+                    // local function, not for the CompilationUnitSyntax. This is desirable in almost call cases but this one:
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:1983]:
```suggestion
                    // local function, not for the CompilationUnitSyntax. This is desirable in almost all cases but this one:
```

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler for a quick review.

**Review by RikkiGibson — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs [hunk 2]
```diff
@@ -1975,7 +1975,24 @@ protected void EnsureNullabilityAnalysisPerformedIfNecessary()
 
             BoundNode bind(CSharpSyntaxNode root, DiagnosticBag diagnosticBag, out Binder binder)
             {
-                binder = GetEnclosingBinder(GetAdjustedNodePosition(root));
+                if (root is CompilationUnitSyntax)
+                {
+                    // Top level statements are unique among our nodes: if there are no syntax nodes before local functions,
+                    // then that means the start of the span of the top-level statement is the same as the start as the local
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:1981]: >as the [](start = 113, length = 6)

"of the"?

File: src/Compilers/CSharp/Test/Symbol/Symbols/Source/NullablePublicAPITests.cs [hunk 3]
```diff
@@ -4712,5 +4712,47 @@ public C()
             Assert.Equal(PublicNullableAnnotation.NotAnnotated, typeInfo.Type.NullableAnnotation);
             Assert.Equal(PublicNullableFlowState.NotNull, typeInfo.Nullability.FlowState);
         }
+
+        [Fact, WorkItem(47467, "https://github.com/dotnet/roslyn/issues/47467")]
+        public void GetDeclaredSymbolTopLevelStatementsWithLocalFunctionFirst_1()
+        {
+            var comp = CreateCompilation(@"void M() {}", options: TestOptions.ReleaseExe.WithNullableContextOptions(NullableContextOptions.Enable));
```
Comment (cston) [inline src/Compilers/CSharp/Test/Symbol/Symbols/Source/NullablePublicAPITests.cs:4719]: >@"void M() {}" [](start = 41, length = 14)

Could this be an `[InlineData()]` argument with the same test method shared across all three tests?

**Review by cston — APPROVED** [review]

**Pull request merged**